config.go

```diff
@@ -159,57 +173,90 @@ func (ca *Cagent) userAgent() string {
 
 func (ca *Cagent) DumpConfigToml() string {
 	buff := &bytes.Buffer{}
-	enc := toml.NewEncoder(buff)
-	err := enc.Encode(ca)
 
+	err := toml.NewEncoder(buff).Encode(ca)
 	if err != nil {
 		log.Errorf("DumpConfigToml error: %s", err.Error())
+		return ""
 	}
 
 	return buff.String()
 }
 
-func (ca *Cagent) ReadConfigFromFile(configFilePath string, createIfNotExists bool) error {
-	dir := filepath.Dir(configFilePath)
-	err := os.MkdirAll(dir, 0755)
+func (ca *Cagent) ReadConfigFromFile(configFilePath string) error {
+	_, err := os.Stat(configFilePath)
 	if err != nil {
-		log.WithError(err).Errorf("Failed to create the config dir: '%s'", dir)
+		return err
 	}
 
-	if _, err := os.Stat(configFilePath); os.IsNotExist(err) {
-		if !createIfNotExists {
-			return fmt.Errorf("Config file not exists: %s", configFilePath)
-		}
-		f, err := os.OpenFile(configFilePath, os.O_WRONLY|os.O_CREATE, 0644)
+	_, err = toml.DecodeFile(configFilePath, &ca)
 
-		if err != nil {
-			return fmt.Errorf("Failed to create the default config file: '%s'", configFilePath)
-		}
-		defer f.Close()
-		enc := toml.NewEncoder(f)
-		enc.Encode(ca)
-	} else {
-		_, err = os.Stat(configFilePath)
-		if err != nil {
-			return err
+	ca.lookupEnv()
+
+	return err
+}
+
+func (ca *Cagent) CreateDefaultConfigFile(configFilePath string) error {
+	var err error
+	defer func() {
+		if err == nil {
+			log.Infof("generated minimum valuable config to %s", configFilePath)
 		}
-		_, err = toml.DecodeFile(configFilePath, &ca)
-		if err != nil {
-			return err
+	}()
+
+	if _, err = os.Stat(configFilePath); os.IsExist(err) {
+		return fmt.Errorf("config already exists at path: %s", configFilePath)
+	}
+
+	var f *os.File
+	if f, err = os.OpenFile(configFilePath, os.O_WRONLY|os.O_CREATE, 0644); err != nil {
+		return fmt.Errorf("failed to create the default config file: '%s'", configFilePath)
+	}
+
+	defer func() {
+		if e := f.Close(); e != nil {
+			log.WithError(e).Errorf("failed to close config file: %s", configFilePath)
 		}
+	}()
+
+	if _, err = f.Write(configAutogeneratedHeadline); err != nil {
+		return fmt.Errorf("failed to write headline to config file")
 	}
 
+	ca.lookupEnv()
```

Comment:
Same as before

Reply:
Agree about rename but it should be here otherwise generated config contains empty values

Reply:
Why not call `lookupEnv()`  before calling this function then?
The more one function does internally the harder it becomes to write isolated tests later

Reply:
Should caller take care of full process of config parse?

```go
if cfgPathPtr != nil {
		err := ca.ReadConfigFromFile(*cfgPathPtr)
		ca.ApplyEnv()
		if os.IsNotExist(err) {
			// this is ok
			err = ca.CreateDefaultConfigFile(*cfgPathPtr)
			if err != nil {
				log.Fatal(err)
			}
		} else if err != nil {
			if strings.Contains(err.Error(), "cannot load TOML value of type int64 into a Go float") {
				log.Fatalf("Config load error: please use numbers with a decimal point for numerical values")
			} else {
				log.Fatalf("Config load error: %s", err.Error())
			}
		}
	}
```

Reply:
Looks better but pay attention to the logical flow.
Why would `ca.ApplyEnv()` only be called in the case of `if cfgPathPtr != nil`
Wouldn't it make more sense to call it _after_ the whole handling of config file is done? Meaning outside the `if` branch

Reply:
It was an example
Looks like ``` if cfgPathPtr != nil { ``` can be removed as by current flow it either provided by user or set as default path otherwise environment must be applied twice (if ```ReadConfigFromFile``` does not guarantee ca fields rewrite in case of error)

Reply:
>Looks like if cfgPathPtr != nil { can be removed as by current flow it either provided by user or set as default path

true

So first try to read config from file and then override with values from env/flags in a defined order. We probably should document the order as well then.